### PR TITLE
Refactor remote message processor

### PR DESCRIPTION
**Pull request by kian-thompson**
The `RemoteMessageProcessor` is really difficult to follow, and making changes to it involves a lot of thinking. This PR changes things to be explicitly called instead of implicitly happening or being skipped behind the various `process` calls.

### CODE REVIEW

File: packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts:32]: This needs a comment, at the very least.  If you can update callsites to be after unpacking, that would be even better IMO.  But I can imagine that may not be trivial.

File: packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts:116]: Took me a minute to recognize this syntax, for some reason the `_` variable name tripped me up.  Maybe `_dropped` instead?  Could be only me...

Comment (markfields) [reply packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts:116]: Or do `type: undefined` in the `return` statement?

File: packages/runtime/container-runtime/src/opLifecycle/opSplitter.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/opSplitter.ts:215]: nit: I would just inline the Final/Notfinal shapes rather than naming them, since they're not used otherwise.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:333]: Adding `disableGroupedBatching` to this `if` statement looks like a bug fix, is that right?  Might want to pull that and any other behavior changes into a separate PR for clarity - so this one can be pure refactor with no behavior change.

If you do that, consider adding all the new test coverage too (if it doesn't depend on the test coverage), to "prove" that the refactor doesn't change behavior.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:406]: Nice, pulling the grouping logic out of here is cool

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:52]: It was intentional to call this `Copy`, to indicate it can be modified without affecting other parts of the system.  See the doc comment above.

File: packages/runtime/container-runtime/src/opLifecycle/opDecompressor.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/opDecompressor.ts:89]: Can you add a brief doc comment here?  When reading the RMP code there's not much context to what "store" might mean - It can be a way someone would begin to learn about how Op Compression/Decompression works

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:63]: This detail is distracting from the overall flow.  Consider putting the undefined check inside `unpackRuntimeMessage`, with a similar comment for the scenario that needs it?

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:57]: Can you add a big fat comment at the top here explaining the various states the incoming op could be in, and how this will deal with all those perumtations?

Also, it's hard for me to recognize how this covers all 6 permutations.  Before I could trace through it with the nesting, but I'm not seeing the symmetry I expected.  Hopefully that comment would help someone like me understand.

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:74]: Chunked messages are always compressed? Could use a comment to that effect.

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:88]: Can you just do this instead?

```suggestion
			return this.opGroupingManager.ungroupOp(message).forEach(unpack);
```

File: packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts:32]: After looking at the usages, it seems like you can know whether it's unpacked yet or not. Take a look, would be better to be explicit/tighter.

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:91]: This is contrary to what I just suggested about `isGroupedBatch`, but it looks like since that and `ungroupOp` are resilient to handle either packed or unpacked messages, I'm curious if you can move this `unpack` up to inside the `if (this.opDecompressor.currentlyUnrolling) {` block above, and ditch the `needToUnpack` flag.

Or, leave it and make those things stricter.

**Review by markfields — COMMENTED** [review]
I didn't review the tests, left some comments.  I can Approve if no one else does, but if someone else has time to look closer at the tests as well that would probably be good.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts
Comment (kian-thompson) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:333]: https://github.com/microsoft/FluidFramework/pull/20163

File: packages/runtime/container-runtime/src/opLifecycle/opDecompressor.ts
Comment (andre4i) [inline packages/runtime/container-runtime/src/opLifecycle/opDecompressor.ts:146]: #nit considering you're returning, there is no use for the `else` statement. This way, the code will look less nested and more like a set of guard clauses.

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (andre4i) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:94]: I think this PR is a great opportunity to document this with some sort of (state transition|flow) diagram (something like https://github.com/microsoft/FluidFramework/pull/17273/files) because even thought it's cleaner and more straight-forward, there are still a lot of possible states here.

**Review by andre4i — COMMENTED** [review]
Looks good on first glance, I will review the tests later

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:94]: To clarify, I don't mean that it needs to be part of the PR, I mean that it's a great opportunity to do it in the near future while the refactoring is still fresh in your mind. Doing it later can be as daunting as someone coming in new to this area of the code 😃

**Review by andre4i — APPROVED** [review]
Thanks for doing this to improve the codebase. I don't have any strong feelings about my comments, but please read through them. Wait for Mark's responses as he had some relevant feedback, otherwise this is 👍

File: packages/runtime/container-runtime/src/test/opLifecycle/remoteMessageProcessor.spec.ts
Comment (kian-thompson) [reply packages/runtime/container-runtime/src/test/opLifecycle/remoteMessageProcessor.spec.ts:226]: I added tests for all 6 combination of these options:
https://github.com/microsoft/FluidFramework/pull/20162/files#diff-12543f825062cd74e55d75a2c446df22de29360d2e4f27142ec5a51b525e03d7R136

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts
Comment (kian-thompson) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:91]: @andre4i Do you have any concerns about this call moving to the bottom? It was originally at the beginning of `RemoteMessageProcessor.process(...)`

**Pull request merged**
